src/music/Instruments.ts

```diff
@@ -0,0 +1,96 @@
+import Instrument from "./Instrument.js";
+import { Envelopes, AdsrConfig } from "./Envelopes.js";
+import { Note, Notes } from "./Notes.js";
+import { SampleData } from "./Samples.js";
+
+type OscType = OscillatorNode["type"];
+
+export interface OscillatorConfig {
+    type: OscType; // either 'sine', 'triangle', 'square', or 'sawtooth'
+    detune?: number; // detune in cents
+}
+
+// A basic instrument that uses a basic waveform (with or without detune).
+export class OscillatorInstrument extends Instrument {
+
+    type: OscType;
+    private _detune: number;
+    env: AdsrConfig;
+
+    constructor(osc: OscillatorConfig, env: AdsrConfig, volume: number = 1) {
+        super(volume);
+        this.type = osc.type;
+        this._detune = 1;
+        this.detune = osc.detune || 0;
+        this.env = env;
+    }
+
+    set detune(n: number) {
+        this._detune = Math.pow(2, n / 1200);
+    }
+
+    scheduleNote(context: AudioContext, note: Note): AudioNode {
+        const freqs: number[] = Notes.detuneWithCoeff(note.midiNumber, this._detune);
+        // note ending frequencies are either where they started or at the endNote
+        const endfreqs: number[] = (note.endNote === undefined) ? freqs : Notes.detuneWithCoeff(note.endNote, this._detune);
+        // volume scaling so multiple oscillators from detune don't result in louder sounds
+        const volume = Math.min(1 / freqs.length, 1) * this.volume;
+        // create envelope
+        const gain: GainNode = Envelopes.createAdsrEnvelope(context, note.start, note.duration,
+            this.env, volume);
+        const end = note.start + note.duration + (this.env.sustain || 0);
+        // initialize oscillator(s)
+        const oscs: OscillatorNode[] = freqs.map((freq, i) => {
+            const osc = context.createOscillator();
+            osc.type = this.type;
+            osc.frequency.setValueAtTime(freq, note.start);
+            osc.frequency.linearRampToValueAtTime(endfreqs[i], end);
+            osc.start(note.start);
+            osc.stop(end);
+            osc.connect(gain);
+            return osc;
+        });
+        oscs[0].onended = () => {
+            gain.disconnect();
+        };
```

Comment:
Since the gain node should really be detached after all of the oscillator nodes have ended, consider instead something like:
```typescript
const oscs: Promise<void>[] = freqs.map((freq, i) => {
    const osc = context.createOscillator();
    osc.type = this.type;
    osc.frequency.setValueAtTime(freq, note.start);
    osc.frequency.linearRampToValueAtTime(endfreqs[i], end);
    osc.start(note.start);
    osc.stop(end);
    osc.connect(gain);
    return new Promise((resolve) => {
        osc.onended = () => resolve();
    });
});
Promise.all(oscs).then(() => gain.disconnect(), () => {});
```